homeassistant/components/nest/__init__.py

```diff
@@ -94,10 +120,128 @@
+    hass.data[DOMAIN] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    if CONF_PROJECT_ID not in config[DOMAIN]:
+        return await async_setup_legacy(hass, config)
+
+    if CONF_SUBSCRIBER_ID not in config[DOMAIN]:
+        _LOGGER.error("Configuration option '{CONF_SUBSCRIBER_ID}' required")
+        return False
+
+    # For setup of ConfigEntry below
+    hass.data[DOMAIN][DATA_NEST_CONFIG] = config[DOMAIN]
+    project_id = config[DOMAIN][CONF_PROJECT_ID]
+    config_flow.NestFlowHandler.register_sdm_api(hass)
+    config_flow.NestFlowHandler.async_register_implementation(
+        hass,
+        config_entry_oauth2_flow.LocalOAuth2Implementation(
+            hass,
+            DOMAIN,
+            config[DOMAIN][CONF_CLIENT_ID],
+            config[DOMAIN][CONF_CLIENT_SECRET],
+            OAUTH2_AUTHORIZE.format(project_id=project_id),
+            OAUTH2_TOKEN,
+        ),
+    )
+
+    return True
+
+
+class SignalUpdateCallback(EventCallback):
+    """An EventCallback invoked when new events arrive from subscriber."""
+
+    def __init__(self, hass: HomeAssistant):
+        """Initialize EventCallback."""
+        self._hass = hass
+
+    def handle_event(self, event_message: EventMessage):
+        """Process an incoming EventMessage."""
+        _LOGGER.debug("Update %s @ %s", event_message.event_id, event_message.timestamp)
+        traits = event_message.resource_update_traits
+        if traits:
+            _LOGGER.debug("Trait update %s", traits.keys())
+        events = event_message.resource_update_events
+        if events:
+            _LOGGER.debug("Event Update %s", events.keys())
+
+        if not event_message.resource_update_traits:
+            # Note: Currently ignoring events like camera motion
+            return
+        # This event triggered an update to a device that changed some
+        # properties which the DeviceManager should already have received.
+        # Send a signal to refresh state of all listening devices.
+        dispatcher_send(self._hass, SIGNAL_NEST_UPDATE)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Nest from a config entry with dispatch between old/new flows."""
+
+    if DATA_SDM not in entry.data:
+        return await async_setup_legacy_entry(hass, entry)
+
+    implementation = (
+        await config_entry_oauth2_flow.async_get_config_entry_implementation(
+            hass, entry
+        )
+    )
+
+    config = hass.data[DOMAIN][DATA_NEST_CONFIG]
+
+    session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)
+    auth = api.AsyncConfigEntryAuth(
+        aiohttp_client.async_get_clientsession(hass),
+        session,
+        API_URL,
+    )
+    subscriber = GoogleNestSubscriber(
+        auth, config[CONF_PROJECT_ID], config[CONF_SUBSCRIBER_ID]
+    )
+    subscriber.set_update_callback(SignalUpdateCallback(hass))
+    hass.async_create_task(subscriber.start_async())
```

Comment:
If this is a long living task, you should use `hass.loop.create_task`. `hass.async_create_task` used during startup of Home Assistant will keep track of the task and will block startup from being finished.
```suggestion
    hass.loop.create_task(subscriber.start_async())
```

Reply:
`SignalUpdateCallback.handle_event` calls `dispatcher_send`, which is a synchronous function. Here I see that we're running the subscriber async. Does the subscriber trigger handle events in sync context?

Reply:
This subscriber is a wrapper around the Google subscriber Client subscribe method:
https://googleapis.dev/python/pubsub/latest/subscriber/api/client.html#google.cloud.pubsub_v1.subscriber.client.Client.subscribe
Note: we are not currently using a scheduler.

So the subscriber is started async which runs a background thread.   It then invokes our callback, which:
(1) updates the state in the DeviceManager and (2) invokes our handle_event  then (3) once that is done, our subscriber calls ack() on the message.

I think we'd be OK with blocking the subscriber thread, and waiting for the response before acking the message.

I copied this dispatcher_send with a signal from the existing nest integration, which has a somewhat similar model.

Here are some options I can see:
1) Might there be a simple way to avoid using signal at all? Each Device could have its own "add_update_callback" that gets invoked directly when the DeviceManager updates its traits or sends it an event (e.g. person_motion, etc).  I think this might be needed anyway once we get to implementing cameras or binary_sensor, but i'm not certain.
2) Switch this all to be async
3) I think we may be OK with blocking the subscriber thread, and ACKing once this returns.  Maybe extra comments explaining the sync vs async model of the whole thing.

Open to other ideas if you have them, or where you'd like to see extra comments.